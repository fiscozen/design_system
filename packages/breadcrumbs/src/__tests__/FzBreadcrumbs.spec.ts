import { describe, it, expect } from 'vitest'
import { mount, flushPromises } from '@vue/test-utils'
import { RouteRecordRaw, createRouter, createWebHistory, useRoute, useRouter } from 'vue-router'

import { Breadcrumb, CustomRouteLocation, FzRouterBreadcrumbs } from '..'

const breadcrumbs: Breadcrumb<CustomRouteLocation>[] = [
  {
    id: 'home',
    label: 'home',
    metadata: {
      name: 'home',
      path: '/'
    }
  },
  {
    id: 'foo',
    label: 'foo',
    metadata: {
      name: 'foo',
      path: '/foo'
    }
  },
  {
    id: 'bar',
    label: 'bar',
    metadata: {
      name: 'bar',
      path: '/foo/bar'
    }
  },
  {
    id: 'baz',
    label: 'baz',
    metadata: {
      name: 'baz',
      path: '/foo/baz'
    }
  }
]

const Page = {
  setup() {
    const route = useRoute()
    const router = useRouter()

    const routes = router.getRoutes()

    return {
      route,
      routes
    }
  },
  template: `
    <div>
        <h2>Page {{ route.name }}</h2>                
        <pre>full path: {{route.fullPath}}</pre>
        <h3>Test links</h3>
        <div class="flex flex-row">
          <router-link class="px-4 underline" v-for="routeRec in routes" :to="routeRec.path">{{routeRec.name}}</router-link>
        </div>
    </div>
  `
}

const routes: RouteRecordRaw[] = [
  {
    name: 'home',
    path: '/',
    component: Page,
    children: [
      {
        name: 'foo',
        path: '/foo',
        component: Page,
        children: [
          {
            name: 'bar',
            path: '/foo/bar',
            component: Page
          }
        ]
      },
      {
        name: 'baz',
        path: '/foo/baz',
        component: Page
      }
    ]
  }
]

const router = createRouter({
  history: createWebHistory(),
  routes
})

const App = {
  components: { FzRouterBreadcrumbs },
  template: `
    <fz-router-breadcrumbs />
    <router-view />
  `
}

describe('FzRouterBreadcrumbs', () => {
  it('should match snapshot for static breadcrumbs', async () => {
    router.push('/')
    await router.isReady()

    const wrapper = mount(FzRouterBreadcrumbs, {
      props: {
        breadcrumbs
      },
      global: {
        plugins: [router]
      }
    })

    await flushPromises()
    expect(wrapper.html()).toMatchSnapshot()
    expect(
      wrapper.findComponent({ name: 'fz-breadcrumbs' }).findAllComponents({ name: 'router-link' })
    ).to.have.length(4)
  })

  it('should match snapshot for automatic breadcrumbs', async () => {
    router.push('/')
    await router.isReady()

    const wrapper = mount(App, {
      global: {
        plugins: [router]
      }
    })

    await router.push('/foo/bar')
    await flushPromises()

    expect(wrapper.html()).toMatchSnapshot()
    expect(
      wrapper
        .findComponent({ name: 'fz-router-breadcrumbs' })
        .findComponent({ name: 'fz-breadcrumbs' })
        .findAllComponents({ name: 'router-link' })
    ).to.have.length(3)
  })
})
