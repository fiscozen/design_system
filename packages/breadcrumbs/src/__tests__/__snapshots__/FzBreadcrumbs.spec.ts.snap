// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FzRouterBreadcrumbs > should match snapshot for automatic breadcrumbs 1`] = `
"<div class="fz__breadcrumbs">
  <div class="flex text-sm"><a href="/" class="router-link-active text-blue-500">home</a>
    <div class="text-grey-300 mx-4">/</div><a href="/foo" class="router-link-active text-blue-500">foo</a>
    <div class="text-grey-300 mx-4">/</div><a aria-current="page" href="/foo/bar" class="router-link-active router-link-exact-active text-blue-500 text-grey-500">bar</a>
    <!--v-if-->
  </div>
</div>
<div>
  <h2>Page bar</h2><pre>full path: /foo/bar</pre>
  <h3>Test links</h3>
  <div class="flex flex-row"><a href="/foo/bar" class="router-link-active router-link-exact-active px-4 underline" aria-current="page">bar</a><a href="/foo/baz" class="px-4 underline">baz</a><a href="/foo" class="router-link-active px-4 underline">foo</a><a href="/" class="router-link-active px-4 underline">home</a></div>
</div>"
`;

exports[`FzRouterBreadcrumbs > should match snapshot for static breadcrumbs 1`] = `
"<div class="fz__breadcrumbs">
  <div class="flex text-sm"><a aria-current="page" href="/" class="router-link-active router-link-exact-active text-blue-500">home</a>
    <div class="text-grey-300 mx-4">/</div><a href="/foo" class="text-blue-500">foo</a>
    <div class="text-grey-300 mx-4">/</div><a href="/foo/bar" class="text-blue-500">bar</a>
    <div class="text-grey-300 mx-4">/</div><a href="/foo/baz" class="text-blue-500 text-grey-500">baz</a>
    <!--v-if-->
  </div>
</div>"
`;
